DAOS-16362: pydaos.torch checkpointing

enakta opened the pull request:
Introducing PyTorch checkpoint interface and user documentation for `pydaos.torch` module.



### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/1/execution/node/133/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/2/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/3/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/4/execution/node/135/log

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15691/3/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15691/3/display/redirect

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15691/3/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15691/3/display/redirect

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/5/execution/node/135/log

daltonbohning commented on docs/user/pytorch.md:
```diff
@@ -0,0 +1,77 @@
+It is widely used in the research community and in the industry.
+PyTroch allows loading data from various sources and DAOS can be used as a storage backend for training data and model's checkpoints.
+
+[DFS plugin](https://github.com/daos-stack/daos/tree/master/src/client/pydaos/torch) implements PyTroch interfaces for loading data from DAOS: Map and Iterable style datasets.
```
```suggestion
[DFS plugin](https://github.com/daos-stack/daos/tree/master/src/client/pydaos/torch) implements PyTorch interfaces for loading data from DAOS: Map and Iterable style datasets.
```

jolivier23 replied:
The main commit message has this same typo

daltonbohning reviewed (commented):
Mostly nits. Overall looks good!

Since `directory_tree.py` was modified we should run the test that uses it. You can do that by including this string in future commit messages
```
Features: DfuseFind
```

mchaarawi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +690,162 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, dir_name, O_RDWR, &dir, NULL, NULL);
```
i have to point out this again that a full path lookup on every write is inefficient :-)

enakta replied:
yup, the upcoming patch should improve that along with the read part

mchaarawi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +690,162 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+	dfs_obj_t         *obj       = NULL;
+	PyObject          *buffer    = NULL;
+	int                oflags    = O_RDWR | O_CREAT | O_TRUNC;
+	mode_t             mode      = S_IFREG | S_IRWXU | S_IRWXG | S_IRWXO;
```
wouldn't the mode be passed down from the torch module somehow? is this the standard permissions and can they be modified by the module?

enakta replied:
Was trying to come up with sane defaults, will change them to be passed from python module.

enakta replied:
done

mchaarawi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +690,162 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+	dfs_obj_t         *dir       = NULL;
+	dfs_obj_t         *obj       = NULL;
+	PyObject          *buffer    = NULL;
+	int                oflags    = O_RDWR | O_CREAT | O_TRUNC;
```
so the behavior you want is truncate an existing file if it exists and overwrite it?

enakta replied:
Yes, the checkpoint is created once per epoch. Truncate is used in case the model is retrained in place, e.g. during benchmarking. Nonetheless, that's a good point, I'll change it to be passed from the python module.

mchaarawi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +690,162 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+		goto out;
+	}
+
+	rc = dfs_open(hdl->dfs, dir, file_name, mode, oflags, 0, 0, NULL, &obj);
```
would be nice to find a way to allow manipulating the object class at least for this call and not use the default. like some env variable or such. if the file is small, uisng the default object class might not be efficient for the file size query.

enakta replied:
makes sense, will do

enakta replied:
done

mchaarawi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +690,162 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+
+	assert(hdl->dfs != NULL);
+
+	int rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, &st);
```
we can optimize this a bit if the dir path is cached to just call dfs_stat. but for now this is fine.

mchaarawi reviewed (dismissed):
the C shim part looks good to me. no changes requested; just some clarifications / comments added.

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/6/execution/node/133/log

enakta commented on src/tests/ftest/directory_tree.py:
```diff
@@ -162,14 +168,32 @@ def _create_needle(self, current_path, current_height):
+        if self._file_size_max > self._file_size_min:
+            size = random.randrange(self._file_size_min, self._file_size_max)
+
+        with os.fdopen(fd, 'wb') as f:
+            f.write(os.urandom(size))
+        return fname
```
`with` calls close after leaving the scope

daltonbohning replied:
I thought that just closes `f`, not `fd`? I'm not sure

daltonbohning replied:
I did check locally that `fd` is in fact closed.

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/7/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/8/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/9/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/10/execution/node/133/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/11/execution/node/134/log

daosbuild1 commented:
Test stage Python Bandit check completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/12/execution/node/134/log

enakta commented:
@mchaarawi I've added parameters passthrough for `dfs_open`.
@johannlombardi I've added parallel, chunked checkpoint writes.

Could you please have another look ?

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/12/execution/node/1184/log

daltonbohning commented on src/tests/ftest/pytorch/dataset.py:
```diff
@@ -0,0 +1,249 @@
+
+        actual = {}
+        for _, content in enumerate(dataset):
+            h = hashlib.md5(content).hexdigest()
```
Python bandit is complaining with
> Use of weak MD5 hash for security. Consider usedforsecurity=False

This is test code so not a real concern, but the easiest way to silence is to resolve
```suggestion
            h = hashlib.md5(content, usedforsecurity=False).hexdigest()
```

daltonbohning commented on src/tests/ftest/pytorch/dataset.py:
```diff
@@ -0,0 +1,249 @@
+
+        actual = {}
+        for _, content in enumerate(dataset):
+            h = hashlib.md5(content).hexdigest()
```
```suggestion
            h = hashlib.md5(content, usedforsecurity=False).hexdigest()
```

daltonbohning commented on src/tests/ftest/pytorch/dataset.py:
```diff
@@ -0,0 +1,249 @@
+        actual = {}
+        for batch in loader:
+            for content in batch:
+                h = hashlib.md5(content).hexdigest()
```
```suggestion
                h = hashlib.md5(content, usedforsecurity=False).hexdigest()
```

daltonbohning commented on src/tests/ftest/directory_tree.py:
```diff
@@ -162,14 +170,32 @@ def _create_needle(self, current_path, current_height):
+
+        size = self._file_size_min
+        if self._file_size_max > self._file_size_min:
+            size = random.randrange(self._file_size_min, self._file_size_max)
```
Python bandit
> Standard pseudo-random generators are not suitable for security/cryptographic purposes.

I think we can ignore this one like
```suggestion
            size = random.randrange(self._file_size_min, self._file_size_max)  # nosec
```

daltonbohning commented on src/tests/ftest/pytorch/checkpoint.py:
```diff
@@ -0,0 +1,100 @@
+
+        writes = []
+        for _ in range(num_writes):
+            writes.append(os.urandom(random.randint(min_size, max_size)))
```
Python bandit. We have a wrapper for this one since it's common :)
```suggestion
            writes.append(os.urandom(self.random.randint(min_size, max_size)))
```

daltonbohning commented on src/tests/ftest/pytorch/checkpoint.py:
```diff
@@ -0,0 +1,100 @@
+
+        writes = []
+        for _ in range(num_writes):
+            writes.append(os.urandom(random.randint(min_size, max_size)))
```
```suggestion
            writes.append(os.urandom(self.random.randint(min_size, max_size)))
```

daltonbohning reviewed (commented):
Python bandit  (security check) has some complaints. Since this is test code it's not a real concern but we should get the check clean

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15691/14/testReport/

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/14/execution/node/1140/log

jolivier23 commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -680,6 +687,170 @@ __shim_handle__torch_batch_read(PyObject *self, PyObject *args)
+
+out:
+	D_FREE(cp1);
+	D_FREE(cp2);
```
D_FREE(*dir) ?

enakta replied:
`*dir` is freed a bit above in case when `name` cannot be allocated. There's no need to free it otherwise, as it's expected output.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/16/execution/node/1220/log

johannlombardi commented on src/client/pydaos/torch/__init__.py:
```diff
@@ -1,32 +1,72 @@
-# (C) Copyright 2024 Google LLC
-# (C) Copyright 2024 Enakta Labs Ltd
+# (C) Copyright 2024-2025 Intel Corporation.
+# (C) Copyright 2025 Hewlett Packard Enterprise Development LP
```
You should probably not touch those.

enakta replied:
Linting was failing without these 🤷‍♂️

jolivier23 replied:
utils/cq/check_update_copyright.sh

Just add handling for Enakta there and you should be fine

jolivier23 replied:
we can still land it with copyright issues though if you just want to change them

johannlombardi replied:
i would change that since the patch need to be updated.

enakta replied:
updated, it fails now requesting copyright in yaml files 😂

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15691/17/testReport/

enakta commented:
Looks like lazy python shim module initialisation does not always work correctly, most likely the same issue as https://github.com/daos-stack/daos/pull/15277.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15691/22/execution/node/1565/log

johannlombardi commented on docs/user/pytorch.md:
```diff
@@ -0,0 +1,77 @@
+# DAOS pytorch interface
```
You need to add an entry for this file to https://github.com/daos-stack/daos/blob/master/mkdocs.yml

johannlombardi commented on src/client/pydaos/torch/torch_shim.c:
```diff
@@ -1,13 +1,15 @@
- * (C) Copyright 2024 Enakta Labs Ltd
+ * (C) Copyright 2024-2025 Google LLC
+ * (C) Copyright 2024-2025 Enakta Labs Ltd
+ * (C) Copyright 2025 Hewlett Packard Enterprise Development LP
```
same comment as before here. Pls just update the Enakta and Google copyright.

johannlombardi commented on src/tests/ftest/directory_tree.py:
```diff
@@ -1,5 +1,7 @@
 """
   (C) Copyright 2018-2024 Intel Corporation.
+  (C) Copyright 2025 Hewlett Packard Enterprise Development LP
+  (C) Copyright 2025 Google LLC
```
same

johannlombardi commented on src/tests/ftest/pytorch/checkpoint.py:
```diff
@@ -0,0 +1,99 @@
+"""
+  (C) Copyright 2025 Hewlett Packard Enterprise Development LP
```
same. you should not add the HPE copyright on this new file.

johannlombardi commented on src/tests/ftest/pytorch/dataset.py:
```diff
@@ -0,0 +1,249 @@
+"""
+  (C) Copyright 2025 Hewlett Packard Enterprise Development LP
```
same

johannlombardi commented on src/client/pydaos/torch/torch_api.py:
```diff
@@ -253,16 +254,263 @@ def __load_batch(self, items):
+                self._offset += len(self._buffer)
+                self._buffer = bytearray()
+
+            data = data[len(chunk):]
```
Is that a very efficient way to copy vs managing it in C with memcpy?

enakta replied:
Haven't notice any performance hits using `bytearray`, but `data = data[len(chunk):]` will create extra copies, indeed.
I've pushed the fix to use `memoryview` instead to avoid copies in the loop.

johannlombardi commented on src/client/pydaos/torch/torch_api.py:
```diff
@@ -253,16 +254,263 @@ def __load_batch(self, items):
+        until some of the chunks are written to the storage.
+        """
+
+        self._queue.put((offset, chunk))
```
could you please elaborate on how the queue size is managed? Is it driven by the upper layer?

enakta replied:
Yes, the queue size can be limited by `chunks_limit` parameter when `Checkpoint` object is created.

johannlombardi reviewed (changes requested):
The copyright must be fixed at the minimum IMO.

mjmac commented:
@daos-stack/daos-gatekeeper: This is ready to be force-landed. All of the jenkins tests passed. Not sure what the linter is complaining about, but I don't think it should block landing.

Recommended message (taken from https://github.com/daos-stack/daos/pull/15814 which was pushed proactively):
```
DAOS-16362 client: pydaos.torch checkpointing

Introducing PyTorch checkpoint interface and user
documentation for pydaos.torch module. Includes
misc. fixes to pydaos.torch module.

Signed-off-by: Denis Barakhtanov <dbarahtanov@enakta.com>
```

mjmac commented:
@daos-stack/daos-gatekeeper: This is ready to be force-landed. All of the jenkins tests passed. Not sure what the linter is complaining about, but I don't think it should block landing.

Recommended message (taken from https://github.com/daos-stack/daos/pull/15814 which was pushed proactively):
```
DAOS-16362 client: pydaos.torch checkpointing

Introducing PyTorch checkpoint interface and user
documentation for pydaos.torch module. Includes
misc. fixes to pydaos.torch module.

Signed-off-by: Denis Barakhtanov <dbarahtanov@enakta.com>
```

daltonbohning commented on src/tests/ftest/pytorch/dataset.yaml:
```diff
@@ -0,0 +1,44 @@
+hosts:
+  test_servers: 1
+  test_clients: 1
+server_config:
```
We need a timeout here. The last run took ~90s so I'd recommend 180 for cushion. I'll see about adding a lint check for this so it's caught early
```suggestion
  test_clients: 1
timeout: 180
server_config:
```

mjmac replied:
I'll fix this specifically in a follow-on if there aren't any other changes coming. Would be great to get this landed as-is so that we don't need to rebase other work.

johannlombardi replied:
@enakta pls push this in another PR.

enakta replied:
@johannlombardi https://github.com/daos-stack/daos/pull/15829 created

daltonbohning reviewed (approved):
Can push the timeout addition in a separate PR if needed. Otherwise LGTM
